plugins/dell-kestrel/fu-dell-kestrel-package.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2024 Dell Technologies
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR MIT
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-dell-kestrel-common.h"
+#include "fu-dell-kestrel-ec-hid.h"
+#include "fu-dell-kestrel-ec-struct.h"
+#include "fu-dell-kestrel-package.h"
+
+#define FU_DELL_KESTREL_PACKAGE_VERSION_OFFSET 0x14
+
+struct _FuDellKestrelPackage {
+	FuDevice parent_instance;
+	guint64 blob_version_offset;
+};
+
+G_DEFINE_TYPE(FuDellKestrelPackage, fu_dell_kestrel_package, FU_TYPE_DEVICE)
+
+static gchar *
+fu_dell_kestrel_package_convert_version(FuDevice *device, guint64 version_raw)
+{
+	return fu_version_from_uint32_hex(version_raw, fu_device_get_version_format(device));
+}
+
+static gboolean
+fu_dell_kestrel_package_setup(FuDevice *device, GError **error)
+{
+	FuDevice *proxy = fu_device_get_proxy(device);
+	guint32 pkg_version_raw;
+	guint8 dock_sku = 0;
+	FuDellDockBaseType dock_type;
+
+	/* instance ID */
+	dock_type = fu_dell_kestrel_ec_get_dock_type(proxy);
+	dock_sku = fu_dell_kestrel_ec_get_dock_sku(proxy);
+
+	fu_device_add_instance_u8(device, "DOCKTYPE", dock_type);
+	fu_device_add_instance_u8(device, "DOCKSKU", dock_sku);
+	fu_device_add_instance_strup(device, "DEVTYPE", "PACKAGE");
+	fu_device_build_instance_id(device, error, "EC", "DOCKTYPE", "DOCKSKU", "DEVTYPE", NULL);
+
+	/* setup version */
+	pkg_version_raw = fu_dell_kestrel_ec_get_package_version(proxy);
+	fu_device_set_version_raw(device, pkg_version_raw);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_kestrel_package_write(FuDevice *device,
+			      FuFirmware *firmware,
+			      FuProgress *progress,
+			      FwupdInstallFlags flags,
+			      GError **error)
+{
+	FuDevice *proxy = fu_device_get_proxy(device);
+	gsize length = 0;
+	guint32 status_version = 0;
+	const guint8 *data;
+	g_autofree gchar *dynamic_version = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	g_return_val_if_fail(device != NULL, FALSE);
+	g_return_val_if_fail(FU_IS_FIRMWARE(firmware), FALSE);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	data = g_bytes_get_data(fw, &length);
+	if (!fu_memcpy_safe((guint8 *)&status_version,
+			    sizeof(status_version),
+			    0x0,
+			    data,
+			    length,
+			    FU_DELL_KESTREL_PACKAGE_VERSION_OFFSET,
+			    sizeof(status_version),
+			    error))
+		return FALSE;
+
+	/* new version */
+	dynamic_version =
+	    fu_version_from_uint32_hex(status_version, fu_device_get_version_format(device));
+
+	g_info("writing firmware: %s, %s -> %s",
+	       fu_device_get_name(device),
+	       fu_device_get_version(device),
+	       dynamic_version);
+
+	if (!fu_dell_kestrel_ec_commit_package(proxy, fw, error))
+		return FALSE;
+
+	/* dock will reboot to re-read; this is to appease the daemon */
+	fu_device_set_version(device, dynamic_version); /* nocheck:set-version */
```

Comment:
didn't you add a new flag for this purpose?  also UoD should it influence this behavior?

Reply:
nah, pkg write to nvm directly.